File: 0_1.baekjoon/graph/java/src/BOJ13023.java
import java.io.*;
import java.util.*;
public class BOJ13023 {
    private static int N, M;
    private static Map<Integer, List<Integer>> graph;
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        graph = new HashMap<>();

        for(int i=0;i<M;i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());

            graph.putIfAbsent(u, new ArrayList<>());
            graph.putIfAbsent(v, new ArrayList<>());

            graph.get(u).add(v);
            graph.get(v).add(u);
        }

        System.out.println(solve());
    }

    private static int solve() {
        for (int i=0;i<N;i++) {
            boolean[] visited = new boolean[N];
            visited[i] = true;
            if (dfs(i, visited, 1)) {
                return 1;
            }
        }
        return 0;
    }

    private static boolean dfs(int node, boolean[] visited, int depth) {
        if (depth == 5) {
            return true;
        }

        for (int adj : graph.getOrDefault(node, new ArrayList<>())) {
            if (visited[adj]) continue;
            visited[adj] = true;
            if (dfs(adj, visited, depth + 1)) {
                return true;
            }
            visited[adj] = false;
        }
        return false;
    }
}

File: 0_1.baekjoon/graph/kotlin/src/BOJ13023.kt
private var N = 0
private var M = 0
private lateinit var graph: HashMap<Int, MutableList<Int>>

fun main() = with(System.`in`.bufferedReader()) {
    val(n, m) = readLine().split(" ").map { it.toInt() }
    N = n
    M = m

    graph = HashMap()

    repeat(m) {
        val (u, v) = readLine().split(" ").map { it.toInt()}
        graph.putIfAbsent(u, mutableListOf())
        graph.putIfAbsent(v, mutableListOf())

        graph[u]!!.add(v)
        graph[v]!!.add(u)
    }

    print(solve())
}

private fun solve(): Int {
    for (i in 0 until N) {
        var visited = BooleanArray(N) { false}
        visited[i] = true
        if(dfs(i, visited, 1)) {
            return 1
        }
    }
    return 0
}

private fun dfs(node: Int, visited: BooleanArray, depth: Int): Boolean {
    if (depth == 5) {
        return true
    }

    for (adj in graph.getOrDefault(node, mutableListOf())) {
        if (visited[adj]) continue
        visited[adj] = true
        if(dfs(adj, visited, depth + 1)) {
            return true
        }
        visited[adj] = false
    }
    return false
}